procrastinate/sql/schema.sql

```diff
@@ -42,55 +42,48 @@ CREATE TABLE procrastinate_events (
     at timestamp with time zone DEFAULT NOW() NULL
 );
 
-CREATE TABLE procrastinate_job_locks (
-    object text PRIMARY KEY
-);
-
 CREATE FUNCTION procrastinate_fetch_job(target_queue_names character varying[]) RETURNS procrastinate_jobs
     LANGUAGE plpgsql
     AS $$
 DECLARE
-	found_jobs procrastinate_jobs;
+    found_job procrastinate_jobs;
 BEGIN
-	WITH potential_job AS (
-		SELECT procrastinate_jobs.*
-			FROM procrastinate_jobs
-			LEFT JOIN procrastinate_job_locks ON procrastinate_job_locks.object = procrastinate_jobs.lock
-			WHERE (target_queue_names IS NULL OR queue_name = ANY( target_queue_names ))
-			  AND procrastinate_job_locks.object IS NULL
-			  AND status = 'todo'
-			  AND (scheduled_at IS NULL OR scheduled_at <= now())
-            ORDER BY id ASC
-			FOR UPDATE OF procrastinate_jobs SKIP LOCKED LIMIT 1
-	), lock_object AS (
-		INSERT INTO procrastinate_job_locks
-			SELECT lock FROM potential_job
-            ON CONFLICT DO NOTHING
-            RETURNING object
-	)
-	UPDATE procrastinate_jobs
-		SET status = 'doing'
-		FROM potential_job, lock_object
-        WHERE lock_object.object IS NOT NULL
-		AND procrastinate_jobs.id = potential_job.id
-		RETURNING procrastinate_jobs.* INTO found_jobs;
-
-	RETURN found_jobs;
+    WITH
+    candidates AS (
+        SELECT DISTINCT ON (lock) id
+        FROM procrastinate_jobs
+        WHERE status IN ('todo', 'doing')
+        ORDER BY lock, id ASC
```

Comment:
defeats the purpose of making the function lockless, no ? This will cost a lot if there is something in the todo/doing

Reply:
Indeed.

I've done a quick test with 50000 jobs in the "todo" state in the procrastinate_jobs table.

With the current master:

```
procrastinate=# select procrastinate_fetch_job('{"default"}');
                   procrastinate_fetch_job                    
--------------------------------------------------------------
 (17,default,a,61a716dafae85a08278e83eba33c4464,,{},doing,,0)
(1 row)

Time: 4.953 ms
procrastinate=# select procrastinate_fetch_job('{"default"}');
                   procrastinate_fetch_job                    
--------------------------------------------------------------
 (18,default,a,e77042cda133e3c709651390de64b64c,,{},doing,,0)
(1 row)

Time: 4.690 ms
procrastinate=# select procrastinate_fetch_job('{"default"}');
                   procrastinate_fetch_job                    
--------------------------------------------------------------
 (19,default,a,92dfc9bec6664bb4181bcc0b2e430807,,{},doing,,0)
(1 row)

Time: 2.871 ms
procrastinate=# select procrastinate_fetch_job('{"default"}');
                   procrastinate_fetch_job                    
--------------------------------------------------------------
 (20,default,a,a5b2f2f0c7e41180d00a02dd4d4ea314,,{},doing,,0)
(1 row)
```

With this PR's branch:

```
procrastinate=# select procrastinate_fetch_job('{"default"}');
                   procrastinate_fetch_job                    
--------------------------------------------------------------
 (17,default,a,f52539d521a24d87f0f23934fba7b029,,{},doing,,0)
(1 row)

Time: 275.783 ms
procrastinate=# select procrastinate_fetch_job('{"default"}');
                   procrastinate_fetch_job                    
--------------------------------------------------------------
 (18,default,a,4ad87374801b611019ee68bdcd3ab839,,{},doing,,0)
(1 row)

Time: 275.847 ms
procrastinate=# select procrastinate_fetch_job('{"default"}');
                   procrastinate_fetch_job                    
--------------------------------------------------------------
 (19,default,a,ef28f468878863eb4e22cb1b129befc7,,{},doing,,0)
(1 row)

Time: 294.295 ms
procrastinate=# select procrastinate_fetch_job('{"default"}');
                   procrastinate_fetch_job                    
--------------------------------------------------------------
 (20,default,a,043d92f2683589933b3751ffe35cb9ad,,{},doing,,0)
(1 row)

Time: 273.440 ms
```